rfcs/2024-05-21-ui-component-model.md

```diff
@@ -0,0 +1,140 @@
+# UI component model
+
+Authors: Gordon Brander
+
+## Overview
+
+Context: converge on a default UI component model for LLM-generated UI.
+
+## Goals
+
+### Product goals
+
+- Easy for LLMs to generate
+    - Leverages patterns and/or frameworks that are widely present in the training data, or can be learned within a small context window.
+- Easy for humans to edit
+    - [Maybe it’s not even code?](https://x.com/threepointone/status/1792930000766677034) Or maybe it’s a hybrid of code and plain language?
+- Leverages familiar or established patterns for UI development
+- Conformable with existing web FE toolchains.
+
+### Technical goals
+
+- Components are **encapsulated** (P1)
+    - A component may control its child tree, but not its siblings or parent
+- Components have **inputs**, and **output UI** and **events** (P1)
+    - Components can be understood as pure-ish functions that receive props and return a view description (we may allow cheats for local component state ala hooks)
+    - Components are **black boxes**
+    - Components are decoupled and only communicate via input and output channels.
+- Component **inputs** and **outputs** are **statically-typed** (P1)
+    - E.g. via TypeScript
+    - Allows the runtime to enforce data policies on component
+- Components have **local state** (P1)
+    - State is encapsulated within component
+    - Components may pass state down to child components as input
+        - E.g. the React **[“lifting state up” pattern](https://legacy.reactjs.org/docs/lifting-state-up.html)**.
+    - Local state may be **persisted**.
+        - If it isn’t, it is **ephemeral**, and lasts for the lifetime of the component.
+- Components are **islands** (P1)
+    - Components can be used free-standing, or within a larger component tree.
+    - Note: this is in contrast to something like the Elm App Architecture Pattern, where models, views, and update functions are “zippered” together, meaning components are “some assembly required”. This would fall short of this goal without some additional means of making an individual component a free-standing island.
+- Components are **composable** (P1)
+    - Components can be combined together like lego to create larger components
+    - Composing plugins should be as easy as plugging together component inputs, outputs, and events, and arranging a UI tree. It shouldn’t be more complicated than that.
```

Comment:
Do you envision composition where we get back one big vdom expression containing all nested children, or something more like Web Components where you register named tags and use them by name? (my expectation is the former, just asking out loud).

Reply:
is composition at the recipe level (i.e. the children vdom is fed into the parent component by the runtime) or within the component (i.e. it's a big black box to the runtime)?

also, we'll want to not allow parents to inspect or change the child vdom for security reasons, but we can implement that later, e.g. by treating the children vdom as a passthrough signal and referencing that in the output vdom. that would also optimize any later vdom updates.